flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/StructuredToStringCastRule.java

```diff
@@ -0,0 +1,249 @@
+                            methodCall(inputTerm, "isNullAt", fieldIndex))
+                    .ifStmt(
+                            "!" + fieldIsNullTerm,
+                            thenBodyWriter ->
+                                    thenBodyWriter
+                                            // If element not null, extract it and
+                                            // execute the cast
+                                            .assignStmt(
+                                                    fieldTerm,
+                                                    CodeGenUtils.rowFieldReadAccess(
+                                                            fieldIndex,
+                                                            inputTerm,
+                                                            attribute.getType()))
+                                            .append(codeBlock)
+                                            .stmt(
+                                                    methodCall(
+                                                            builderTerm,
+                                                            "append",
+                                                            strLiteral(attribute.getName() + "=")
+                                                                    + " + "
+                                                                    + codeBlock.getReturnTerm())),
+                            elseBodyWriter ->
+                                    // If element is null, just write NULL
+                                    elseBodyWriter.stmt(
+                                            methodCall(
+                                                    builderTerm,
+                                                    "append",
+                                                    strLiteral(attribute.getName() + "=")
+                                                            + " + "
+                                                            + nullLiteral(
+                                                                    context.legacyBehaviour()))));
+        }
+
+        writer.stmt(methodCall(builderTerm, "append", strLiteral("}")));
+
+        return CharVarCharTrimPadCastRule.padAndTrimStringIfNeeded(
+                        writer,
+                        targetLogicalType,
+                        context.legacyBehaviour(),
+                        length,
+                        resultStringTerm,
+                        builderTerm)
+                // Assign the result value
+                .assignStmt(
+                        returnVariable,
+                        CastRuleUtils.staticCall(
+                                BINARY_STRING_DATA_FROM_STRING(), resultStringTerm))
+                .toString();
+    }
+
+    private static String getDelimiter(CodeGeneratorCastRule.Context context) {
+        final String comma;
+        if (context.legacyBehaviour()) {
```

Comment:
isn't the `legacyBehaviour` actually to use the row representation instead of map?

Reply:
TBH I don't think there is any legacy behaviour for this casting at all, as users weren't able to execute `CAST(structuredVal AS STRING)` anyway.

Reply:
Are you sure about this? I thought we used the ROW representation for casting in Flink 1.14?

Reply:
Yes, we have the code in the runtime, but `LogicalTypeChecks` (still) doesn't validate the cast types tuple